RecoParticleFlow/PFProducer/plugins/importers/TrackTimingImporter.cc

```diff
@@ -0,0 +1,89 @@
+#include "DataFormats/ParticleFlowReco/interface/PFRecTrackFwd.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecTrack.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/MuonReco/interface/MuonFwd.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "RecoParticleFlow/PFProducer/interface/PFMuonAlgo.h"
+#include "RecoParticleFlow/PFTracking/interface/PFTrackAlgoTools.h"
+
+// this doesn't actually import anything, 
+// but rather applies time stamps to tracks after they are all inserted
+
+class TrackTimingImporter : public BlockElementImporterBase {
+public:
+  TrackTimingImporter(const edm::ParameterSet& conf,
+		    edm::ConsumesCollector& sumes) :
+    BlockElementImporterBase(conf,sumes),    
+    srcTime_( sumes.consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("timeValueMap")) ),
+    srcTimeError_( sumes.consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("timeErrorMap")) ),
+    srcTimeGsf_( sumes.consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("timeValueMapGsf")) ),
+    srcTimeErrorGsf_( sumes.consumes<edm::ValueMap<float> >(conf.getParameter<edm::InputTag>("timeErrorMapGsf")) ),
+    debug_(conf.getUntrackedParameter<bool>("debug",false)) {    
+  }
+  
+  void importToBlock( const edm::Event& ,
+		      ElementList& ) const override;
+
+private:
+    
+  edm::EDGetTokenT<edm::ValueMap<float> > srcTime_, srcTimeError_, srcTimeGsf_, srcTimeErrorGsf_;
+  const bool debug_;
+  
+};
+
+DEFINE_EDM_PLUGIN(BlockElementImporterFactory, 
+		  TrackTimingImporter, 
+		  "TrackTimingImporter");
+
+void TrackTimingImporter::
+importToBlock( const edm::Event& e, 
+	       BlockElementImporterBase::ElementList& elems ) const {
+  typedef BlockElementImporterBase::ElementList::value_type ElementType;  
+  
+  edm::Handle<edm::ValueMap<float> > timeH, timeErrH, timeGsfH, timeErrGsfH;
+  
+  e.getByToken(srcTime_, timeH);
+  e.getByToken(srcTimeError_, timeErrH);
+  e.getByToken(srcTimeGsf_, timeGsfH);
+  e.getByToken(srcTimeErrorGsf_, timeErrGsfH);
+  
+  auto TKs_end = std::partition(elems.begin(),elems.end(),
+				[](const ElementType& a){
+			        return a->type() == reco::PFBlockElement::TRACK;
+				});  
+  auto btk_elems = elems.begin();  
+  for( auto track = btk_elems;  track != TKs_end; ++track) {
+    const auto& ref = (*track)->trackRef();
+    if (timeH->contains(ref.id())) {
+      (*track)->setTime( (*timeH)[ref], (*timeErrH)[ref] );
+    }
+    if( debug_ ) {
+      edm::LogInfo("TrackTimingImporter") 
+      	<< "Track with pT / eta " << ref->pt() << " / " << ref->eta() 
+	<< " has time: " << (*track)->time() << " +/- " << (*track)->timeError() << std::endl;
+    }
+    
+  }
+  
+  auto gsfTKs_end = std::partition(elems.begin(),elems.end(),
+				[](const ElementType& a){
+			        return a->type() == reco::PFBlockElement::GSF;
+				});  
```

Comment:
related to observed changes in reco: this and the above reorder the elements in blocks.
IIRC, PF reco is not invariant under element reordering.
Maybe stable_partition instead?

Reply:
Don't think stable partition will help since this is doing a global resort of all elements.
Perhaps a better way to do this is to not use partition and loop through all elements and alter them in place?

Reply:
Longer execution time, but should be no physics impact even from element reordering then.

Reply:
is there any significant effect expected in having all first elements be GSF tracks, followed by CKF tracks, and then everything else?
If it's a matter of a random permutation, it may be OK. Still, it's a change in behavior where none is really anticipated.

Reply:
Is there a canonical/conventional ordering of the block elements in general?

Reply:
My guess is that the original order of importers sets the initial order.

Reply:
just to be clear: I'm now expecting a fix to this reordering issue.
[the discussion yesterday was pointing in the direction that it's about to happen already]